### PR TITLE
Check refresh token flow response for offline based on refresh token …

### LINKED ISSUE #30857: Check for being Offline type in refresh token flow must be done based on refresh token request parameter
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

oidc

### Describe the bug

Based on [Refreshing an Access Token](https://datatracker.ietf.org/doc/html/rfc6749#section-6) : 

> The authorization server MAY issue a new refresh token, in which case
   the client MUST discard the old refresh token and replace it with the
   new refresh token.  The authorization server MAY revoke the old
   refresh token after issuing a new refresh token to the client. 

In Keycloak you can not return refresh token in refresh token flow using `SuppressRefreshTokenRotationExecutor` policy executor.
Another way is after geting refresh token to disable 'Use refresh tokens' client parameter.

User get refresh token for a client  'Use refresh tokens' including `offline_access` scope. This is offline type refresh token.
User do refresh token flow. A new access token will be returned without a new refresh token.

[This PR](https://github.com/keycloak/keycloak/pull/12196) has added check for being offline type based on return refresh token (if exists).
As you see from above scenario, this could lead to an offline refresh flow to characterize as no offline.
This lead in [RefreshTokenGrantType](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/oidc/grants/RefreshTokenGrantType.java) line 85 to try update user session for offline user session.

For regular cases, although no error is thrown, I do not know what happens during offline user session update.
**The most important is that this scenario will not work during Keycloak restart (or upgrade). After restart, if you try to execute refresh token flow with a refresh token produced before restart, a NullPointerException will be thrown .As you know, offline refresh token lives after restart**

**Check for refresh token must be done based on refresh token request parameter.**

### Version

25.0.1

### Regression

- [X] The issue is a regression

### Expected behavior

- In the scenario mentioned, user session must not refreshed in refresh token flow.

- Offline refresh token flow must always work with offline refresh token produced before Keycloak restart.

### Actual behavior

- In the scenario mentioned, user session is refreshed in refresh token flow.

- Offline refresh token flow with offline refresh token produced before Keycloak restart, returns NullPointerException.

### How to Reproduce?

- Client with 'Use refresh tokens' parameter.
- Use `SuppressRefreshTokenRotationExecutor` policy executor for this client.
- User get refresh token for this client  including `offline_access` scope. This is offline type refresh token.
- Keycloak restart
- Execute refresh token flow with a refresh token produced before restart

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by cgeorgilakis**
…request parameter

Closes #30857

### CODE REVIEW

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExtendedEventTest.java [hunk 1]
```diff
@@ -486,6 +486,54 @@ public void testExtendedClientPolicyIntefacesForTokenRefreshResponse() throws Ex
...
+        json = (new ClientPoliciesBuilder()).addPolicy(
+                (new ClientPolicyBuilder()).createPolicy(POLICY_NAME, "Den Forste Politikken", Boolean.TRUE)
+                        .addCondition(ClientRolesConditionFactory.PROVIDER_ID,
+                                createClientRolesConditionConfig(Arrays.asList(SAMPLE_CLIENT_ROLE)))
+                        .addProfile(PROFILE_NAME)
+                        .toRepresentation()
+        ).toString();
+        updatePolicies(json);
+
+        String refreshTokenString = res.getRefreshToken();
```
Comment (rmartinc) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExtendedEventTest.java:531]: You need to force the NPE deleting the non-offline session before doing the refresh. Something like:

```diff
diff --git a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExtendedEventTest.java b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExtendedEventTest.java
index 16922436ac..5a73efec6e 100644
--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExtendedEventTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExtendedEventTest.java
@@ -509,6 +509,8 @@ public class ClientPoliciesExtendedEventTest extends AbstractClientPoliciesTest
 
         OAuthClient.AccessTokenResponse res = oauth.doAccessTokenRequest(code, clientSecret);
         assertEquals(200, res.getStatusCode());
+        AccessToken token = oauth.verifyToken(res.getAccessToken());
+        assertNotNull(token.getSessionId());
 
         // register profiles
         String json = (new ClientProfilesBuilder()).addProfile(
@@ -528,6 +530,9 @@ public class ClientPoliciesExtendedEventTest extends AbstractClientPoliciesTest
         ).toString();
         updatePolicies(json);
 
+        // delete the non-offline session to force the NPE
+        adminClient.realm(REALM_NAME).deleteSession(token.getSessionId(), false);
+
         String refreshTokenString = res.getRefreshToken();
         OAuthClient.AccessTokenResponse accessTokenResponseRefreshed = oauth.doRefreshTokenRequest(refreshTokenString, clientSecret);
         assertEquals(200, accessTokenResponseRefreshed.getStatusCode());
```

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExtendedEventTest.java [hunk 2]
```diff
@@ -486,6 +486,54 @@ public void testExtendedClientPolicyIntefacesForTokenRefreshResponse() throws Ex
...
+        OAuthClient.AccessTokenResponse accessTokenResponseRefreshed = oauth.doRefreshTokenRequest(refreshTokenString, clientSecret);
+        assertEquals(200, accessTokenResponseRefreshed.getStatusCode());
+        assertEquals(null, accessTokenResponseRefreshed.getRefreshToken());
```
Comment (rmartinc) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExtendedEventTest.java:534]:
```suggestion
        assertNull(accessTokenResponseRefreshed.getRefreshToken());
```

Just a nit-pick. :smile:

**Review by rmartinc — CHANGES_REQUESTED** [review]
Thanks @cgeorgilakis! The change looks good to me but the test does not fail without the changes. You need to delete the non-offline session that is generated before the refresh. With the proposed changes I think I could generate the NPE and the error. Please test it and incorporate to the PR.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExtendedEventTest.java (hunk 1 above)
Comment (cgeorgilakis) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExtendedEventTest.java:531]: Thanks for the help.
Althouth this is the right test, I do not know how to do it in test.

Comment (rmartinc) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExtendedEventTest.java:531]: I made it fail with those lines. Just check it and if it also seems OK to you just add the lines.

**Review by rmartinc — APPROVED** [review]
Thanks @cgeorgilakis! LGTM now!

**Comment by cgeorgilakis** [comment]
> Thanks @cgeorgilakis! LGTM now!

You could see the error with previous implementation [here](https://github.com/eosc-kc/keycloak/actions/runs/9760571359)

**Review by rmartinc — APPROVED** [review]
Yep! I tested it locally and it failed for me too. Thanks!

**Review by pedroigor — APPROVED** [review]

**Pull request merged**